src/os/bluestore/BlueFS.cc

```diff
@@ -3038,54 +3036,67 @@ void BlueFS::_consume_dirty(uint64_t seq)
+  }
+  // inject new allocation in case log is too big
+  if (log.t.expected_size() > runway) {
+    _extend_log(log.t.expected_size(), ignore_compaction);
+  }
+  runway = log.writer->file->fnode.get_allocated() - log.writer->get_effective_write_pos();
+  return runway;
+}
+
+int64_t BlueFS::_extend_log(uint64_t amount, bool ignore_compaction) {
   ceph_assert(ceph_mutex_is_locked(log.lock));
-  // allocate some more space (before we run out)?
-  // BTW: this triggers `flush()` in the `page_aligned_appender` of `log.writer`.
-  int64_t runway = log.writer->file->fnode.get_allocated() -
-    log.writer->get_effective_write_pos();
-  if (runway < (int64_t)cct->_conf->bluefs_min_log_runway) {
-    dout(10) << __func__ << " allocating more log runway (0x"
-	     << std::hex << runway << std::dec  << " remaining)" << dendl;
-    /*
-     * Usually, when we are low on space in log, we just allocate new extent,
-     * put update op(log) to log and we are fine.
-     * Problem - it interferes with log compaction:
-     * New log produced in compaction will include - as last op - jump into some offset (anchor) of current log.
-     * It is assumed that log region (anchor - end) will contain all changes made by bluefs since
-     * full state capture into new log.
-     * Putting log update into (anchor - end) region is illegal, because any update there must be compatible with
-     * both logs, but old log is different then new log.
-     *
-     * Possible solutions:
-     * - stall extending log until we finish compacting and switch log (CURRENT)
-     * - re-run compaction with more runway for old log
-     * - add OP_FILE_ADDEXT that adds extent; will be compatible with both logs
-     */
-    if (log_forbidden_to_expand.load() == true) {
-      return -EWOULDBLOCK;
+  if (!ignore_compaction) {
+    std::unique_lock<ceph::mutex> ll(log.lock, std::adopt_lock);
+    while (log_forbidden_to_expand.load() == true) {
+      log_cond.wait(ll);
     }
-    vselector->sub_usage(log.writer->file->vselector_hint, log.writer->file->fnode);
+    ll.release();
+  }
+  vselector->sub_usage(log.writer->file->vselector_hint, log.writer->file->fnode);
+  while (amount > 0) {
     int r = _allocate(
-      vselector->select_prefer_bdev(log.writer->file->vselector_hint),
-      cct->_conf->bluefs_max_log_runway,
-      0,
-      &log.writer->file->fnode);
+        vselector->select_prefer_bdev(log.writer->file->vselector_hint),
+        cct->_conf->bluefs_max_log_runway,
+        0,
+        &log.writer->file->fnode);
+    amount -= std::min(cct->_conf->bluefs_max_log_runway, amount);
     ceph_assert(r == 0);
-    vselector->add_usage(log.writer->file->vselector_hint, log.writer->file->fnode);
-    log.t.op_file_update_inc(log.writer->file->fnode);
   }
-  return runway;
+  vselector->add_usage(log.writer->file->vselector_hint, log.writer->file->fnode);
+
+  bluefs_transaction_t log_extend_transaction;
+  log_extend_transaction.seq = log.t.seq;
```

Comment:
The injected transaction seem to be ok.
I do not see a progression of seq (see `_log_advance_seq`).
Did not check it, but I have a strong suspicion that the next (the original) transaction that will come after it will have same seq.

Reply:
done